internal/ingest/ingest.go

```diff
@@ -0,0 +1,238 @@
+func (i *legIngester) Subscribe(ctx context.Context, p peer.ID) error {
+	sctx, cancel := context.WithCancel(ctx)
+	s, err := i.newPeerSubscriber(sctx, p)
+	if err != nil {
+		cancel()
+		return err
+	}
+
+	// If already subscribed do nothing.
+	if s.watcher != nil {
+		cancel()
+		return nil
+	}
+
+	var cncl context.CancelFunc
+	s.watcher, cncl = s.ls.OnChange()
+	s.cncl = s.cancelFunc(cncl, cancel)
+
+	// Listen updates persist latestSync when sync is done.
+	go i.listenUpdates(ctx, s)
+	return nil
+}
+
+func (i *legIngester) listenUpdates(ctx context.Context, s *sub) {
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		// Persist the latest sync
+		case c := <-s.watcher:
+			err := i.putLatestSync(s.p, c)
+			if err != nil {
+				log.Errorf("Error persisting latest sync: %w", err)
+			}
+		}
+	}
+}
+
+// Unsubscribe to stop listening to advertisement from a specific provider.
+func (i *legIngester) Unsubscribe(ctx context.Context, p peer.ID) error {
+	i.sublk.Lock(p)
+	defer i.sublk.Unlock(p)
+	// Close subscriber
+	i.subs[p].ls.Close(ctx)
+	// Run cancel
+	i.subs[p].cncl()
+	// Delete from map
+	delete(i.subs, p)
+
+	return nil
+}
+
+// Creates a new subscriber for a peer according to its latest sync.
+func (i *legIngester) newPeerSubscriber(ctx context.Context, p peer.ID) (*sub, error) {
+	i.sublk.Lock(p)
+	defer i.sublk.Unlock(p)
+	s, ok := i.subs[p]
+	// If there is already a subscriber for the peer, do nothing.
+	if ok {
+		return s, nil
+	}
+
+	// See if we already synced with this peer.
+	c, err := i.getLatestSync(p)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: Make a request to provider to see if it has any new advertisement
+	// and sync before initializing subscriber?
+
+	// If not synced start a brand new subscriber
+	if c == cid.Undef {
+		ls, err := legs.NewSubscriber(ctx, i.ds, i.host,
+			i.gs, i.subTopic, legs.FilterPeerPolicy(p))
+		if err != nil {
+			return nil, err
+		}
+		s = &sub{p: p, ls: ls}
+		i.subs[p] = s
+		return s, nil
+	}
+	// If yes, start a partially synced subscriber.
+	ls, err := legs.NewSubscriberPartiallySynced(ctx, i.ds, i.host,
+		i.gs, i.subTopic, legs.FilterPeerPolicy(p), c)
+	if err != nil {
+		return nil, err
+	}
+	s = &sub{p: p, ls: ls}
+	i.subs[p] = s
+	return s, nil
+}
+
+func (i *legIngester) Close(ctx context.Context) error {
+	// Unsubscribe from all peers
+	for k := range i.subs {
+		i.Unsubscribe(ctx, k)
+	}
+	return nil
+}
+
+// Get the latest cid synced for the peer.
+func (i *legIngester) getLatestSync(p peer.ID) (cid.Cid, error) {
+	b, err := i.ds.Get(datastore.NewKey(syncPrefix + p.String()))
+	if err != nil {
+		if err == datastore.ErrNotFound {
+			return cid.Undef, nil
+		}
+		return cid.Undef, err
+	}
+	_, c, err := cid.CidFromBytes(b)
+	return c, err
+}
+
+// Tracks latest sync for a specific peer.
+func (i *legIngester) putLatestSync(p peer.ID, c cid.Cid) error {
+	return i.ds.Put(datastore.NewKey(syncPrefix+p.String()), c.Bytes())
```

Comment:
```suggestion
	if !c.Defined() {
		return nil
	}

	return i.ds.Put(datastore.NewKey(syncPrefix+p.String()), c.Bytes())
```
I think you want to ignore undefined cids because otherwise you get a huge number of them and will try to store them.  Put a debug print there and see.  Is that what should be expected?